### PR TITLE
[+] CORE : Add hooks for URL rewriting modules

### PR DESCRIPTION
**Pull request by firstred**
## Description

Adds hooks for those who are brave enough to change PrestaShop's URL schema.
It should provide a toolkit that is extensive enough to radically change PrestaShop's URLs without the need for overrides.
## Steps to Test this Fix

Step 1. Take a rewriting module
Step 2. Migrate the overrides to the hooks
Step 3. Should get the same results

### CODE REVIEW

**Comment by Soullivaneuh** [comment]
Interesting. What is the difference between this and `hookModuleRoutes` (see example on #5342)?

Do you have some concrete use cases?

Could we manage route translations with your addition?

**Comment by firstred** [comment]
- This is made especially for modules that would like to influence general URLs, for e.g. products, categories, CMS pages, etc.
- It allows modules to directly influence the default routes and loaded routes for every language and the requested shop. The direct access makes sure that modules can take and replace whatever they want.
- Yes, route translations can be managed with this addition. It's strange that PrestaShop, by default, doesn't allow different URL schemes for languages (except for the rewrite part), isn't it?

**Comment by Soullivaneuh** [comment]
> Yes, route translations can be managed with this addition. It's strange that PrestaShop, by default, doesn't allow different URL schemes for languages (except for the rewrite part), isn't it?

I agree for **modules routes translations** part.

But for the rest, what is the goal of this? Do you have a concrete example?

What would be the interested to overload core routes generation by module? This already exists on the SEO admin section of Prestashop AFAIK.

**Comment by firstred** [comment]
> But for the rest, what is the goal of this? Do you have a concrete example?


Very simple use case: remove IDs

> What would be the interested to overload core routes generation by module? This already exists on the SEO admin section of Prestashop AFAIK.


It is totally unrelated to the SEO admin section of PrestaShop, because this would allow a module to directly influence the URLs and provide its own admin section, thereby replacing PrestaShop's default SEO.

**Comment by firstred** [comment]
Rebased

File: classes/Link.php [hunk 1]
```diff
@@ -214,6 +299,25 @@ public function getDownQuantityCartURL(
         $id_product_attribute,
         $id_customization = null
     ) {
+        $down_quantity_cart_urls = Hook::exec(
+            'actionGetDownQuanityCartURL',
```
Comment (Shudrum) [inline classes/Link.php]: Hey, there is a typo here :p

**Comment by Shudrum** [comment]
This is a lot of hooks for a very specific use case … But I like it. I will do some tests before merging it, sorry.

**Comment by firstred** [comment]
Yep, I want to add a lot of hooks to PrestaShop and figured that the best way is to go for specific use cases in every PR.

I'm glad that you like it as I also wanted to know if PRs like these are welcome :)

I'm also trying to pass as many variables by reference as possible and would like to encourage the dev team to do so as well, because my observation is that overrides are mainly used to alter existing variables.

**Comment by kpodemski** [comment]
Interesting :+1:

**Comment by Shudrum** [comment]
Passing variables by reference is not a good thing. The problem is that we should not pass simple variables but objects.

Right now, I understand why doing this, maybe the hooks should return false in case of error or the altered array in case of success …

**Comment by firstred** [comment]
Rebased, updated to latest standards and fixed typos.

**Comment by Shudrum** [comment]
Hi @firstred!

I just tried your changes, well… just by refreshing the catalog page on the classic theme with the demo products.

I'm worried about one thing:

capture d ecran 2016-05-25 a 00 28 35

It add a total of 428 `Hook::exec` calls (yes, so many… and I don't really know why).
And the performances drop is really huge…

Maybe we should wait (and I cannot say when it comes) that we use the Symfony events…

**Pull request closed without merging**